xilem_core/src/views/map_state.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use core::marker::PhantomData;
+
+use crate::{DynMessage, MessageResult, Mut, View, ViewId, ViewPathTracker};
+
+/// A view that "extracts" state from a [`View<ParentState,_,_>`] to [`View<ChildState,_,_>`].
+/// This allows modularization of views based on their state.
+pub struct MapState<ParentState, ChildState, V, F = fn(&mut ParentState) -> &mut ChildState> {
+    f: F,
+    child: V,
+    phantom: PhantomData<fn() -> (ParentState, ChildState)>,
+}
+
+/// A view that "extracts" state from a [`View<ParentState,_,_>`] to [`View<ChildState,_,_>`].
+/// This allows modularization of views based on their state.
+///
+/// # Examples
+///
+/// (From the Xilem implementation)
+///
+/// ```ignore
+/// #[derive(Default)]
+/// struct AppState {
+///     count: i32,
+///     other: i32,
+/// }
+///
+/// fn count_view(count: i32) -> impl WidgetView<i32> {
+///     flex((
+///         label(format!("count: {}", count)),
+///         button("+", |count| *count += 1),
+///         button("-", |count| *count -= 1),
+///     ))
+/// }
+///
+/// fn main() -> Result<(), EventLoopError> {
+///     Xilem::new(AppState::default(), |state| {
+///         map_state(count_view(state.count), |state: &mut AppState|  &mut state.count)
+///     })
+///     .run_windowed(EventLoop::with_user_event(), "Map action example".into())?;
```

Comment:
This API is almost certain to change; I can easily imagine this becoming outdated and not being caught - I'd probably just remove this whole `fn main`, and start from an "app_logic" function